### DISCUSSION TITLE
Support for `tox` and `src` directory structure?

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#2718] [Q&A]
@pyvista/developers

1. Should we look into using `tox`/`detox` to test and support different python versions? Or is `tox`/`detox` not typically used anymore in favor of github-actions matrix builds (i.e. our [testing-and-deployment.yml](https://github.com/pyvista/pyvista/blob/main/.github/workflows/testing-and-deployment.yml)). It seems MacOs supports 3.7 and 3.8, Linux and Windows 3.7 - 3.9, but not yet support 3.10. Should this distinction be made more clearly in the `README.rst`?

2. `pytest` recommends a `src/` directory layout. Though I don't like this aesthetially, strong arguments for its usage are made (though there's [a new `--import-mode=importlib` feature for `pytest`](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#test-discovery) now):

- [Cristian Maries, Ionel. "Packaging a python library". May 25, 2014.](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure)
- [Schlawack, Hynek. "Testing and Packaging". Jan 04, 2021](https://hynek.me/articles/testing-packaging/)

I know [we like editable installs](https://github.com/pyvista/pyvista/pull/1977#issuecomment-1002711613), but [packaging pitfalls](https://blog.ionelmc.ro/2014/06/25/python-packaging-pitfalls/) seems to indicate this can cause us not to test against installed code.

**TL;DR**

1. Is there a reason we don't use `tox`?
2. Is there a reason we don't use a `src/` directory format?

**Comment by akaszynski** [comment]
> 1. Should we look into using `tox`/`detox` to test and support different python versions? Or is `tox`/`detox` not typically used anymore in favor of github-actions matrix builds (i.e. our [testing-and-deployment.yml](https://github.com/pyvista/pyvista/blob/main/.github/workflows/testing-and-deployment.yml)).

To me, that's the biggest limitation to using or not using `tox`. That, and the inertia of moving our logic outside of GitHub Actions into `tox`. However, there are workarounds and PyAnsys is using a GitHub action matrix in [pyansys/ansys-templates](https://github.com/pyansys/ansys-templates).

This might take some time to implement and to me it doesn't seem like a high enough priority to warrant a testing refactor.

> 2\. `pytest` recommends a `src/` directory layout. Though I don't like this aesthetially, strong arguments for its usage are made (though there's [a new `--import-mode=importlib` feature for `pytest`](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#test-discovery) now):

I think this is sufficiently "low hanging fruit" to warrant a refactor, and it's also compatible with an editable install.

I originally wanted to do this when converting over to `pyproject.toml` when using `flit`, which BTW supports editable installs. I still think we don't need to move over to adopting PEP 518 quite yet, but IMO `pyproject.toml` is starting to reach critical mass in the Python community. However, moving to `src` is a step along the way and one that I think is worth the investment.
